### DISCUSSION TITLE
Oculus Quest VR Support

### DISCUSSION THREAD

**Discussion by erikpa1** [tauri-apps/tauri#4395] [Ideas]
Hi guys, 

at first, you have done incredible work here, thanks. And now my topic:

I'm 3D web developer using React, Fiber and Three JS , all are amazing technologies for web 3D development (for applications I develop they are much better choice than unity or unreal engine, I don't need performance,...), **BUT**, I can't run them on oculus quest as native app. Android itself is possible via apache cordova (also not the best but only solution on market) but here is oculus quest is dying, because you cant activate XR rendering.

My question here is: could new android version of tauri also support XR core on oculus quest (and desktop maybe too) ? In my opinion integrating this could be really game changer for javascript XR applications and multiplatform development itself too.

(Yes, there are some react project for react vr, but they are trash because they work only with certain versions, which are not supported anymore and some packages I need require latest nodejs.... which will break another packages,.... )

**Comment by saucesome** [comment]
I second this! :) Great topic.

**Comment by yuripourre** [comment]
@erikpa1 I couldn't agree more. Do you have any updates on this? Have you tried recently? Does it work now?

**Reply by erikpa1** [reply]
@yuripourre I haven't tried, but I think it's time to do so. I hope I will get in touch with it till end of the day. But I don't expect it will be working, because when I tried Android WebView from Java, it wasn't working either. There is probably darker magic in oculus quest browser,...

**Reply by yuripourre** [reply]
@erikpa1 please let me know how it goes. Yes, I read that webview doesn't have all the features as the full-fledged browser (including WebXR). I am trying some crazy ideas, I tell you if I make it work.

**Reply by polvallverdu** [reply]
I don't think webview uses the meta quest browser. I didn't really try much, but for WebXR to work the implementation must be direct and injected, but webview seems like a really basic browser.

**Reply by yuripourre** [reply]
@polvallverdu I don't know about internal implementation of Webview in Oculus but I've heard that Google is working in a component called [WebLayer](https://chromium.googlesource.com/chromium/src/+/refs/tags/80.0.3947.0/weblayer/) that supposedly will replace WebView.

In regards to Oculus browser, yesterday I've made some progress but far from ideal. I created an APK that opens the Oculus Browser in VR mode using Intent (Native Android class). It switches to VR mode (Black Screen) but the browser opens in 2D mode I couldn't trigger the WebXR from there yet.

I could successfully upload the APK to Oculus' App Lab (after tweaking the manifest) so maybe it's promising.

This is the farthest I could get:

```java
Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
intent.setComponent(new ComponentName("com.oculus.browser","com.oculus.browser.WebVRActivity"));
this.startActivity(intent);
```

Please let me know if anyone has any other ideas.

**Reply by erikpa1** [reply]
I tried the same, I don't think it will work. We have to reconcile fact that, the solution will be not possible, because Oculus browser have standalone implementation of WebVR. 

I have a little idea, not promising it will maybe work. Running https server service as background process of oculus (or maybe minimalized app) and setting intent url to that server... but if this will somehow work you will probably have warning "unsecure SSL certificate" - this can be maybe manager by installing certificates on device.

**Reply by yuripourre** [reply]
@erikpa1 would you mind clarifying a little more? What this server would do?

I am thinking about compiling chromium as an APK that would work as a launcher so it can wrap any web page. I don't know if it would work with WebXR though.

**Reply by erikpa1** [reply]
Server should serve files from device, thats what server do,… compiling chromium is good idea but question is if android build support full vr experience,… (not only cardboard VR)

**Comment by yuripourre** [comment]
@erikpa1 

> Server should serve files from device, thats what server do,… 

I am not sure if I get it, the APK would be the client and request files to the server that would be serving the files and running on background, right? I don't understand how the VR mode could be invoked.

>compiling chromium is good idea but question is if android build support full vr experience,… (not only cardboard VR)

Yeah, I don't know if it would work. I am assuming that Oculus Browser is also a fork of chromium and also linked to OpenXR. So it might work out-of-the-box: https://chromium.googlesource.com/chromium/src/+/HEAD/device/vr/openxr/

**Reply by erikpa1** [reply]
I don't understand how can you be so  brave that you are going to build chromium on oculus, but you don't understand what I say :D :D  The solution I'm thinking about is easy as it sounds. Imagine it on desktop,  you run command line app which runs your app in https, and you enter "localhost:3000" url to your browser - same I want to try for oculus.

**Reply by yuripourre** [reply]
@erikpa1 haha! Probably because English is not my native language.

Btw, I could build the Chromium APK but I need to modify the manifest to upload to Oculus, I will probably have some news next month. Let's see.

I understood your idea now but... How can it run on Oculus? You need to upload an APK that renders the page that comes from the server right?
